File: skills/sci-plot/assets/case-packs/rf-0178/plot.R
frame_files <- vapply(sys.frames(), function(x) {
  value <- x$ofile
  if (is.null(value)) "" else as.character(value)
}, character(1))
frame_files <- frame_files[nzchar(frame_files)]
if (length(frame_files) > 0) {
  script_path <- normalizePath(frame_files[length(frame_files)], mustWork = TRUE)
} else {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) == 0) stop("Cannot resolve plot.R location")
  script_path <- normalizePath(gsub("~+~", " ", sub("^--file=", "", file_arg[1]), fixed = TRUE), mustWork = TRUE)
}
base_dir <- dirname(script_path)
data_path <- file.path(base_dir, "data.csv")
output_path <- file.path(base_dir, "plot_r.png")
d <- read.csv(data_path, check.names = FALSE, stringsAsFactors = FALSE, fileEncoding = "UTF-8-BOM")
required <- c("completeness", "number_of_cds", "type", "length")
if (!all(required %in% names(d))) stop("data.csv does not contain the expected columns")
if (nrow(d) == 0) stop("data.csv contains no observations")
if (any(!is.finite(d$length)) || any(!is.finite(d$number_of_cds)) || any(!is.finite(d$completeness))) stop("data.csv contains non-finite numeric values")
if (any(d$length <= 0) || any(d$number_of_cds <= 0)) stop("length and number_of_cds must be positive")
groups <- c("WGS", "MAG", "SAG")
if (any(!d$type %in% groups)) stop("data.csv contains an unexpected genome type")
colors <- c(WGS = "#374955", MAG = "#8C6BB1", SAG = "#247BA0")
x <- log10(d$length)
y <- log10(d$number_of_cds)
model <- lm(y ~ x)
slope <- unname(coef(model)[2])
intercept <- unname(coef(model)[1])
fitted_values <- fitted(model)
residual_values <- resid(model)
r_squared <- summary(model)$r.squared
x_grid <- seq(min(x), max(x), length.out = 500)
prediction <- predict(model, newdata = data.frame(x = x_grid), interval = "confidence", level = 0.95)
pairs <- list(c("WGS", "MAG"), c("WGS", "SAG"), c("MAG", "SAG"))
raw_p <- vapply(pairs, function(pair) {
  wilcox.test(
    residual_values[d$type == pair[1]],
    residual_values[d$type == pair[2]],
    alternative = "two.sided",
    exact = FALSE,
    correct = TRUE
  )$p.value
}, numeric(1))
adjusted_p <- p.adjust(raw_p, method = "holm")
effect_sizes <- vapply(pairs, function(pair) {
  first_values <- residual_values[d$type == pair[1]]
  second_values <- residual_values[d$type == pair[2]]
  combined_ranks <- rank(c(first_values, second_values))
  first_u <- sum(combined_ranks[seq_along(first_values)]) - length(first_values) * (length(first_values) + 1) / 2
  2 * first_u / (length(first_values) * length(second_values)) - 1
}, numeric(1))
residual_span <- diff(range(residual_values))
residual_limits <- range(residual_values) + c(-0.035, 0.035) * residual_span
scientific <- function(value) {
  text <- formatC(value, format = "e", digits = 2)
  text <- sub("e-0", "e−", text, fixed = TRUE)
  text <- sub("e-", "e−", text, fixed = TRUE)
  text
}
alpha_color <- function(color, alpha) grDevices::adjustcolor(color, alpha.f = alpha)
format_log <- function(values) {
  result <- format(values, scientific = FALSE, trim = TRUE, big.mark = ",")
  result
}
device_arguments <- list(filename = output_path, width = 16, height = 12.5, units = "in", res = 360, bg = "#F7F5F1")
if (.Platform$OS.type != "windows" && capabilities("cairo")) device_arguments$type <- "cairo"
do.call(png, device_arguments)
layout(
  matrix(c(1, 1, 2, 3, 3, 4, 5, 5, 6), nrow = 3, byrow = TRUE),
  widths = c(5.5, 5.5, 2.25),
  heights = c(1.15, 5.4, 2.65)
)
par(family = "sans", oma = c(5.4, 1.2, 5.0, 1.2), xaxs = "i", yaxs = "i")
x_breaks <- seq(min(x), max(x), length.out = 30)
y_breaks <- seq(min(y), max(y), length.out = 30)
x_histograms <- lapply(groups, function(name) hist(x[d$type == name], breaks = x_breaks, plot = FALSE))
y_histograms <- lapply(groups, function(name) hist(y[d$type == name], breaks = y_breaks, plot = FALSE))
par(mar = c(0.4, 5.2, 0.8, 0.5))
plot(NA, xlim = range(x), ylim = c(0, max(vapply(x_histograms, function(h) max(h$density), numeric(1))) * 1.08), axes = FALSE, xlab = "", ylab = "")
axis(2, las = 1, cex.axis = 0.75, col.axis = "#4E5A60", col = "#8C969A", tck = -0.02)
mtext("Density", side = 2, line = 3.6, cex = 0.78, col = "#34434B")
abline(h = pretty(c(0, max(vapply(x_histograms, function(h) max(h$density), numeric(1))))), col = "#E3E0DA", lwd = 0.7)
for (i in seq_along(groups)) {
  h <- x_histograms[[i]]
  rect(h$breaks[-length(h$breaks)], 0, h$breaks[-1], h$density, col = alpha_color(colors[groups[i]], 0.10), border = NA)
  lines(rep(h$breaks, each = 2)[-c(1, 2 * length(h$breaks))], rep(h$density, each = 2), col = colors[groups[i]], lwd = 1.5)
}
text(min(x) + 0.01 * diff(range(x)), par("usr")[4] * 0.88, "Marginal distributions", adj = c(0, 0.5), cex = 0.86, font = 2, col = "#34434B")
box(bty = "l", col = "#8C969A")
par(mar = c(0.4, 0.6, 0.8, 0.4))
plot.new()
legend(
  "center",
  legend = c(
    sprintf("WGS  n = %s", format(sum(d$type == "WGS"), big.mark = ",")),
    sprintf("MAG  n = %s", format(sum(d$type == "MAG"), big.mark = ",")),
    sprintf("SAG  n = %s", format(sum(d$type == "SAG"), big.mark = ",")),
    "Global OLS fit"
  ),
  pch = c(16, 16, 16, NA),
  lty = c(NA, NA, NA, 1),
  col = c(colors, "#9C3D3A"),
  pt.cex = 1.1,
  lwd = c(NA, NA, NA, 2.2),
  bty = "n",
  xjust = 0,
  cex = 0.82,
  y.intersp = 1.4
)
par(mar = c(4.8, 5.2, 0.5, 0.5))
plot(
  d$length,
  d$number_of_cds,
  type = "n",
  log = "xy",
  xlab = "Genome length (bp)",
  ylab = "Number of coding sequences (CDS)",
  xaxt = "n",
  yaxt = "n",
  cex.lab = 1.03,
  col.lab = "#25333B"
)
x_ticks <- c(5e5, 1e6, 2e6, 5e6, 1e7)
y_ticks <- c(500, 1000, 2000, 5000, 10000)
axis(1, at = x_ticks, labels = format_log(x_ticks), cex.axis = 0.78, col.axis = "#46545A", col = "#8C969A", tck = -0.02)
axis(2, at = y_ticks, labels = format_log(y_ticks), las = 1, cex.axis = 0.78, col.axis = "#46545A", col = "#8C969A", tck = -0.02)
abline(v = x_ticks, h = y_ticks, col = "#D7D5CF", lwd = 0.7)
polygon(
  c(10^x_grid, rev(10^x_grid)),
  c(10^prediction[, "lwr"], rev(10^prediction[, "upr"])),
  border = NA,
  col = alpha_color("#B44E4A", 0.14)
)
for (name in groups) {
  selected <- d$type == name
  points(d$length[selected], d$number_of_cds[selected], pch = 16, cex = 0.42, col = alpha_color(colors[name], 0.38))
}
lines(10^x_grid, 10^prediction[, "fit"], col = "#9C3D3A", lwd = 2.2)
legend(
  "topleft",
  legend = c(
    "Global descriptive OLS on log10 scale",
    sprintf("slope = %.4f   R² = %.4f   n = %s", slope, r_squared, format(nrow(d), big.mark = ","))
  ),
  bty = "o",
  box.col = "#D8D5CE",
  bg = alpha_color("#FFFFFF", 0.94),
  text.col = "#25333B",
  cex = c(0.82, 0.79),
  text.font = c(2, 1),
  x.intersp = 0.3,
  inset = 0.015
)
text(max(d$length) / 1.03, min(d$number_of_cds) * 1.07, "Shaded band: 95% CI for fitted mean on log10 scale", adj = c(1, 0), cex = 0.70, col = "#5A6469")
box(col = "#8C969A")
par(mar = c(4.8, 0.5, 0.5, 0.8))
max_density <- max(vapply(y_histograms, function(h) max(h$density), numeric(1)))
plot(NA, xlim = c(0, max_density * 1.08), ylim = range(y), axes = FALSE, xlab = "Density", ylab = "")
axis(1, cex.axis = 0.72, col.axis = "#4E5A60", col = "#8C969A", tck = -0.02)
abline(v = pretty(c(0, max_density)), col = "#E3E0DA", lwd = 0.7)
for (i in seq_along(groups)) {
  h <- y_histograms[[i]]
  rect(0, h$breaks[-length(h$breaks)], h$density, h$breaks[-1], col = alpha_color(colors[groups[i]], 0.10), border = NA)
  lines(rep(h$density, each = 2), rep(h$breaks, each = 2)[-c(1, 2 * length(h$breaks))], col = colors[groups[i]], lwd = 1.5)
}
box(bty = "l", col = "#8C969A")
par(mar = c(4.9, 5.2, 2.0, 0.5))
residual_list <- lapply(groups, function(name) residual_values[d$type == name])
box_statistics <- vapply(residual_list, function(values) {
  quartiles <- quantile(values, c(0.25, 0.50, 0.75), type = 7, names = FALSE)
  spread <- quartiles[3] - quartiles[1]
  lower <- min(values[values >= quartiles[1] - 1.5 * spread])
  upper <- max(values[values <= quartiles[3] + 1.5 * spread])
  c(lower, quartiles, upper)
}, numeric(5))
box_object <- list(
  stats = box_statistics,
  n = vapply(residual_list, length, integer(1)),
  conf = matrix(NA_real_, nrow = 2, ncol = length(groups)),
  out = numeric(0),
  group = integer(0),
  names = groups
)
bxp(
  box_object,
  horizontal = TRUE,
  outline = FALSE,
  boxfill = alpha_color(colors, 0.30),
  border = "#66747A",
  medcol = "#17252C",
  medlwd = 1.8,
  whiskcol = "#66747A",
  staplecol = "#66747A",
  boxlwd = 1.0,
  whisklwd = 1.1,
  staplelwd = 1.1,
  xlab = "Residual: observed − fitted log10(CDS)",
  ylim = residual_limits,
  las = 1,
  cex.axis = 0.82,
  cex.lab = 0.92,
  col.lab = "#25333B"
)
set.seed(20260727)
for (i in seq_along(groups)) {
  values <- residual_list[[i]]
  points(values, i + runif(length(values), -0.18, 0.18), pch = 16, cex = 0.32, col = alpha_color(colors[groups[i]], 0.14))
}
abline(v = 0, col = "#9C3D3A", lwd = 1.4, lty = 2)
grid(nx = NULL, ny = NA, col = "#E1DED8", lwd = 0.7)
box(bty = "l", col = "#8C969A")
mtext("Residual distributions by genome type", side = 3, line = 0.7, adj = 0, cex = 1.02, font = 2, col = "#25333B")
par(mar = c(4.9, 0.6, 2.0, 0.4))
plot.new()
median_text <- paste(
  sprintf("%s   median %+.4f", groups, vapply(residual_list, median, numeric(1))),
  collapse = "\n"
)
comparison_text <- paste(
  vapply(
    seq_along(pairs),
    function(i) sprintf(
      "%s–%s   Holm p = %s   δ = %+.3f",
      pairs[[i]][1],
      pairs[[i]][2],
      scientific(adjusted_p[i]),
      effect_sizes[i]
    ),
    character(1)
  ),
  collapse = "\n"
)
text(0, 0.98, "Residual summary", adj = c(0, 1), cex = 0.94, font = 2, col = "#25333B")
text(0, 0.80, median_text, adj = c(0, 1), cex = 0.80, col = "#3E4A50", family = "mono")
text(0, 0.49, "Exploratory pairwise tests", adj = c(0, 1), cex = 0.84, font = 2, col = "#25333B")
text(0, 0.36, comparison_text, adj = c(0, 1), cex = 0.73, col = "#3E4A50", family = "mono")
text(0, 0.02, "Two-sided Mann–Whitney U tests; Holm adjustment.\nδ: rank-biserial effect, first group minus second.", adj = c(0, 0), cex = 0.65, col = "#687277")
mtext("Genome length and coding-sequence abundance", side = 3, outer = TRUE, line = 2.6, adj = 0.04, cex = 1.75, font = 2, col = "#1F2D34")
mtext("A global log–log relationship with group-resolved marginal and residual distributions", side = 3, outer = TRUE, line = 0.9, adj = 0.04, cex = 0.98, col = "#59666C")
mtext(
  "Analysis note. Each row is treated as one independent observation. The global OLS model uses genome length as the sole predictor;",
  side = 1,
  outer = TRUE,
  line = 2.1,
  adj = 0.04,
  cex = 0.72,
  col = "#5D676C"
)
mtext(
  "the completeness column is retained but not modeled. The classical OLS CI assumes homoscedastic residuals; residual tests are exploratory.",
  side = 1,
  outer = TRUE,
  line = 3.3,
  adj = 0.04,
  cex = 0.72,
  col = "#5D676C"
)
invisible(dev.off())

File: skills/sci-plot/assets/case-packs/rf-0178/plot.py
from pathlib import Path
import csv
import numpy as np
import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
from matplotlib.ticker import LogLocator, FuncFormatter
from scipy import stats


BASE_DIR = Path(__file__).resolve().parent
DATA_PATH = BASE_DIR / "data.csv"
OUTPUT_PATH = BASE_DIR / "plot_python.png"
GROUPS = ("WGS", "MAG", "SAG")
COLORS = {"WGS": "#374955", "MAG": "#8C6BB1", "SAG": "#247BA0"}


def read_data(path):
    length = []
    cds = []
    completeness = []
    group = []
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        expected = {"completeness", "number_of_cds", "type", "length"}
        if reader.fieldnames is None or not expected.issubset(set(reader.fieldnames)):
            raise ValueError("data.csv does not contain the expected columns")
        for row in reader:
            length.append(float(row["length"]))
            cds.append(float(row["number_of_cds"]))
            completeness.append(float(row["completeness"]))
            group.append(row["type"].strip())
    length = np.asarray(length, dtype=float)
    cds = np.asarray(cds, dtype=float)
    completeness = np.asarray(completeness, dtype=float)
    group = np.asarray(group, dtype=str)
    if len(length) == 0:
        raise ValueError("data.csv contains no observations")
    if not np.all(np.isfinite(length)) or not np.all(np.isfinite(cds)) or not np.all(np.isfinite(completeness)):
        raise ValueError("data.csv contains non-finite numeric values")
    if np.any(length <= 0) or np.any(cds <= 0):
        raise ValueError("length and number_of_cds must be positive")
    unexpected = sorted(set(group) - set(GROUPS))
    if unexpected:
        raise ValueError(f"Unexpected genome types: {unexpected}")
    return length, cds, completeness, group


def holm_adjust(p_values):
    p_values = np.asarray(p_values, dtype=float)
    order = np.argsort(p_values)
    adjusted_sorted = np.maximum.accumulate((len(p_values) - np.arange(len(p_values))) * p_values[order])
    adjusted = np.empty_like(adjusted_sorted)
    adjusted[order] = np.minimum(adjusted_sorted, 1.0)
    return adjusted


def scientific(value):
    return f"{value:.2e}".replace("e-0", "e−").replace("e+0", "e+").replace("e-", "e−")


length, cds, completeness, group = read_data(DATA_PATH)
x = np.log10(length)
y = np.log10(cds)
fit = stats.linregress(x, y)
fitted = fit.intercept + fit.slope * x
residual = y - fitted
n = len(x)
x_grid = np.linspace(x.min(), x.max(), 500)
y_grid = fit.intercept + fit.slope * x_grid
sse = np.sum(residual ** 2)
sxx = np.sum((x - x.mean()) ** 2)
sigma2 = sse / (n - 2)
se_mean = np.sqrt(sigma2 * (1 / n + (x_grid - x.mean()) ** 2 / sxx))
t_critical = stats.t.ppf(0.975, n - 2)
lower = y_grid - t_critical * se_mean
upper = y_grid + t_critical * se_mean
pairs = (("WGS", "MAG"), ("WGS", "SAG"), ("MAG", "SAG"))
raw_p = []
effect_sizes = []
for first, second in pairs:
    first_values = residual[group == first]
    second_values = residual[group == second]
    result = stats.mannwhitneyu(first_values, second_values, alternative="two-sided")
    raw_p.append(result.pvalue)
    effect_sizes.append(2 * result.statistic / (len(first_values) * len(second_values)) - 1)
adjusted_p = holm_adjust(raw_p)
residual_span = residual.max() - residual.min()
residual_limits = (residual.min() - 0.035 * residual_span, residual.max() + 0.035 * residual_span)

plt.rcParams.update(
    {
        "font.family": "DejaVu Sans",
        "font.size": 11,
        "axes.labelsize": 12,
        "axes.titlesize": 13,
        "axes.linewidth": 0.8,
        "xtick.labelsize": 10,
        "ytick.labelsize": 10,
        "legend.fontsize": 10,
        "figure.facecolor": "#F7F5F1",
        "axes.facecolor": "#FFFFFF",
        "savefig.facecolor": "#F7F5F1",
    }
)

fig = plt.figure(figsize=(16, 12.5), dpi=360)
grid = fig.add_gridspec(
    3,
    3,
    width_ratios=(5.5, 5.5, 2.25),
    height_ratios=(1.15, 5.4, 2.65),
    left=0.075,
    right=0.965,
    bottom=0.125,
    top=0.89,
    wspace=0.08,
    hspace=0.18,
)
ax_main = fig.add_subplot(grid[1, :2])
ax_top = fig.add_subplot(grid[0, :2])
ax_summary = fig.add_subplot(grid[0, 2])
ax_right = fig.add_subplot(grid[1, 2])
ax_residual = fig.add_subplot(grid[2, :2])
ax_stats = fig.add_subplot(grid[2, 2])

x_bins = np.linspace(x.min(), x.max(), 30)
y_bins = np.linspace(y.min(), y.max(), 30)
for name in GROUPS:
    selected = group == name
    ax_top.hist(
        x[selected],
        bins=x_bins,
        density=True,
        histtype="stepfilled",
        color=COLORS[name],
        alpha=0.10,
        linewidth=0,
    )
    ax_top.hist(
        x[selected],
        bins=x_bins,
        density=True,
        histtype="step",
        color=COLORS[name],
        alpha=0.95,
        linewidth=1.35,
    )
    ax_right.hist(
        y[selected],
        bins=y_bins,
        density=True,
        histtype="stepfilled",
        orientation="horizontal",
        color=COLORS[name],
        alpha=0.10,
        linewidth=0,
    )
    ax_right.hist(
        y[selected],
        bins=y_bins,
        density=True,
        histtype="step",
        orientation="horizontal",
        color=COLORS[name],
        alpha=0.95,
        linewidth=1.35,
    )
    ax_main.scatter(
        length[selected],
        cds[selected],
        s=12,
        color=COLORS[name],
        alpha=0.38,
        edgecolors="none",
        rasterized=True,
        zorder=2,
    )

ax_main.fill_between(10 ** x_grid, 10 ** lower, 10 ** upper, color="#B44E4A", alpha=0.14, linewidth=0, zorder=1)
ax_main.plot(10 ** x_grid, 10 ** y_grid, color="#9C3D3A", linewidth=2.2, zorder=4)
ax_main.set_xscale("log")
ax_main.set_yscale("log")
ax_main.set_xlabel("Genome length (bp)", labelpad=9)
ax_main.set_ylabel("Number of coding sequences (CDS)", labelpad=9)
ax_main.grid(which="major", color="#D7D5CF", linewidth=0.65, alpha=0.65)
ax_main.grid(which="minor", color="#EAE8E3", linewidth=0.4, alpha=0.45)
ax_main.set_axisbelow(True)
ax_main.xaxis.set_major_locator(LogLocator(base=10, numticks=7))
ax_main.yaxis.set_major_locator(LogLocator(base=10, numticks=7))
ax_main.text(
    0.025,
    0.965,
    f"Global descriptive OLS on log₁₀ scale\nslope = {fit.slope:.4f}   R² = {fit.rvalue ** 2:.4f}   n = {n:,}",
    transform=ax_main.transAxes,
    ha="left",
    va="top",
    fontsize=11,
    color="#25333B",
    bbox={"boxstyle": "round,pad=0.55", "facecolor": "#FFFFFF", "edgecolor": "#D8D5CE", "alpha": 0.94},
    zorder=6,
)
ax_main.text(
    0.975,
    0.035,
    "Shaded band: 95% confidence interval for the fitted mean on the log₁₀ scale",
    transform=ax_main.transAxes,
    ha="right",
    va="bottom",
    fontsize=9.5,
    color="#5A6469",
)

ax_top.set_ylabel("Density", fontsize=10)
ax_top.set_xlim(x.min(), x.max())
ax_top.tick_params(axis="x", which="both", bottom=False, labelbottom=False)
ax_top.tick_params(axis="y", labelsize=9)
ax_top.spines["top"].set_visible(False)
ax_top.spines["right"].set_visible(False)
ax_top.spines["bottom"].set_visible(False)
ax_top.grid(axis="y", color="#E3E0DA", linewidth=0.55)
ax_top.text(0.01, 0.88, "Marginal distributions", transform=ax_top.transAxes, fontsize=10.5, weight="bold", color="#34434B")

ax_right.set_xlabel("Density", fontsize=10)
ax_right.set_ylim(y.min(), y.max())
ax_right.tick_params(axis="y", which="both", left=False, labelleft=False)
ax_right.tick_params(axis="x", labelsize=9)
ax_right.spines["top"].set_visible(False)
ax_right.spines["right"].set_visible(False)
ax_right.spines["left"].set_visible(False)
ax_right.grid(axis="x", color="#E3E0DA", linewidth=0.55)

ax_summary.axis("off")
legend_handles = [
    Line2D([0], [0], marker="o", linestyle="", markerfacecolor=COLORS[name], markeredgecolor="none", markersize=7, label=f"{name}  n = {np.sum(group == name):,}")
    for name in GROUPS
]
legend_handles.append(Line2D([0], [0], color="#9C3D3A", linewidth=2.2, label="Global OLS fit"))
ax_summary.legend(handles=legend_handles, loc="center left", frameon=False, handlelength=2.1, borderaxespad=0)

box_data = [residual[group == name] for name in GROUPS]
box = ax_residual.boxplot(
    box_data,
    vert=False,
    positions=np.arange(1, len(GROUPS) + 1),
    widths=0.48,
    showfliers=False,
    patch_artist=True,
    medianprops={"color": "#17252C", "linewidth": 1.8},
    whiskerprops={"color": "#66747A", "linewidth": 1.1},
    capprops={"color": "#66747A", "linewidth": 1.1},
    boxprops={"edgecolor": "#66747A", "linewidth": 1.0},
)
for patch, name in zip(box["boxes"], GROUPS):
    patch.set_facecolor(COLORS[name])
    patch.set_alpha(0.30)
rng = np.random.default_rng(20260727)
for position, name in enumerate(GROUPS, start=1):
    values = residual[group == name]
    jitter = rng.uniform(-0.18, 0.18, size=len(values))
    ax_residual.scatter(values, position + jitter, s=7, color=COLORS[name], alpha=0.14, edgecolors="none", rasterized=True)
ax_residual.axvline(0, color="#9C3D3A", linewidth=1.4, linestyle=(0, (4, 3)))
ax_residual.set_xlim(residual_limits)
ax_residual.set_yticks(np.arange(1, len(GROUPS) + 1), GROUPS)
ax_residual.set_xlabel("Residual: observed − fitted log₁₀(CDS)", labelpad=8)
ax_residual.set_title("Residual distributions by genome type", loc="left", pad=10, weight="bold", color="#25333B")
ax_residual.grid(axis="x", color="#E1DED8", linewidth=0.6)
ax_residual.set_axisbelow(True)
ax_residual.spines["top"].set_visible(False)
ax_residual.spines["right"].set_visible(False)

ax_stats.axis("off")
median_lines = "\n".join(
    f"{name}   median {np.median(residual[group == name]):+.4f}" for name in GROUPS
)
comparison_lines = "\n".join(
    f"{first}–{second}   Holm p = {scientific(value)}   δ = {effect:+.3f}"
    for (first, second), value, effect in zip(pairs, adjusted_p, effect_sizes)
)
ax_stats.text(0.0, 0.97, "Residual summary", ha="left", va="top", fontsize=11.5, weight="bold", color="#25333B")
ax_stats.text(0.0, 0.79, median_lines, ha="left", va="top", fontsize=10, color="#3E4A50", linespacing=1.65)
ax_stats.text(0.0, 0.47, "Exploratory pairwise tests", ha="left", va="top", fontsize=10.5, weight="bold", color="#25333B")
ax_stats.text(0.0, 0.34, comparison_lines, ha="left", va="top", fontsize=9.8, color="#3E4A50", linespacing=1.65)
ax_stats.text(
    0.0,
    0.01,
    "Two-sided Mann–Whitney U tests; Holm adjustment.\nδ: rank-biserial effect, first group minus second.",
    ha="left",
    va="bottom",
    fontsize=8.9,
    color="#687277",
    linespacing=1.45,
)

fig.suptitle(
    "Genome length and coding-sequence abundance",
    x=0.075,
    y=0.955,
    ha="left",
    fontsize=23,
    weight="bold",
    color="#1F2D34",
)
fig.text(
    0.075,
    0.918,
    "A global log–log relationship with group-resolved marginal and residual distributions",
    ha="left",
    va="center",
    fontsize=12.5,
    color="#59666C",
)
fig.text(
    0.075,
    0.055,
    "Analysis note. Each row is treated as one independent observation. The global OLS model uses genome length as the sole predictor; "
    "the completeness column is retained in the source data but is not included in this model. The classical OLS confidence interval assumes "
    "homoscedastic residuals. Pairwise residual tests are exploratory and do not establish causality.",
    ha="left",
    va="bottom",
    fontsize=9.5,
    color="#5D676C",
    wrap=True,
)
fig.savefig(OUTPUT_PATH, dpi=360, bbox_inches="tight", pad_inches=0.18)
plt.close(fig)
